### DISCUSSION TITLE
suggetion for the wiki entry on GUIs

### DISCUSSION THREAD

**Discussion by JamesWidman** [neovim/neovim#24998] [General]
Hi all!

I'm having difficulty finding a GUI for macOS that supports the use of modifier keys in key mappings.

(this is frustrating, because the ability to use modifier keys like `command` and `option` (aka `meta`, and implicitly option-as-meta) is the only reason why i wanted to use a GUI in the first place; otherwise, i would be using nvim in a terminal.)

I suspect I'm not the only one who has found this to be difficult, so I'm wondering if we could add a new column in the [gui section of the "related projects" wiki page](https://github.com/neovim/neovim/wiki/Related-projects#gui). 

For example, for a GUI that supports the use of all four modifiers (command, option, shift, control), the entry in this new column would read something like, "cmd+opt+shift+control" or maybe even "⌘⌥⇧⌃", whereas for a GUI that supports all but the option key, its entry for this new column would be "⌘⇧⌃").

Or there could be a separate boolean column for each modifier.

Unfortunately, this is complicated by the fact that modifiers are different on different platforms (e.g. "command" is not present on windows or linux, and "alt" on a windows PC keyboard is in a different location than "alt" on a mac keyboard).

One way to resolve this would be to have a separate table for each of linux, windows, and mac. (This might be nice even without the "modifier support" column, since it's never completely clear whether a "cross-platform" GUI actually works on a particular platform. For a GUI that actually works on all three platforms, we would see a little redundancy, but in exchange we would have a little more clarity.)

**Comment by justinmk** [comment]
A good start would be to provide such a list here. Then that can be copied to the wiki.

**Comment by fredizzimo** [comment]
One problem with such a list is that it easily gets out of date, unless it's maintained by the maintainers of the GUIs themselves.

Take for example Neovide, which I'm one of the maintainers of. For the longest time we were limited by the library we used for dealing with the keyboard inputs, Winit, which had a mult-year effort of adding proper support for all platforms. Now when they finally released a new alpha version, we have been able to start fixing the remaining issues, with the aim to support everything perfectly. I guess, other GUIs have similar aim, so anything missing or broken is mostly caused by bugs.

Anyway, to illustrate the situation in Neovide (the situation in other GUs is most likely very similar).

We do support `option-as-meta`, or `neovide_input_macos_alt_is_meta` as we call it. But I think that our implementation is still a bit buggy in certain situations. Additionally, it only applies to both meta keys or none. So currently we have a PR open that will add support for changing the behaviour of only the left or the right key. That one was done in a slightly suboptimal way and is bit stalled, but there's another PR in progress. So, if that was to be part of the wiki it would change very soon.

Simple columns like that would also not tell much. For example, currently the combination CMD+shift is broken in Neovide, while CMD alone works fine. Many CTRL+shift combnations are also broken as a regression. Again, there's a PR open to fix both of those problems. But it won't fix all CMD+shift combinations, we also need a winit fix to fix it all.

Then CTRL+alt gr is also not working properly, but that's a winit issue, so I don't know when it will be fixed, but hopefully it won't take too long.

I hope that this didn't sound too much like an advertisement for Neovide, I just wanted to illustrate the complexity of the situation, and how things rapidly change, and I don't know much about the other GUIs to provide examples from those. 

But I agree that a proper comparison between different GUIs is missing, and it's not only limited to keyboard input, but there are also a lot of other factors to consider and there are a lot of great GUIs out there. Unfortunately, all of them, including Neovide also have their shortcomings. But perhaps that's better done by some independent reviewer than creating a fully objective list in the Neovim wiki?
